pkg/morph/subscriber/subscriber.go

```diff
@@ -199,16 +163,161 @@ func New(ctx context.Context, p *Params) (Subscriber, error) {
+	}
+	// Worker listens all events from temporary NeoGo channel and puts them
+	// into corresponding permanent channels.
 	go sub.routeNotifications(ctx)
 
 	return sub, nil
 }
 
+func (s *subscriber) routeNotifications(ctx context.Context) {
+	var (
+		// TODO: not needed after nspcc-dev/neo-go#2980.
+		cliCh             = s.client.NotificationChannel()
+		restoreCh         chan bool
+		restoreInProgress bool
+	)
+
+routeloop:
+	for {
+		var connLost bool
+		s.RLock()
+		notifCh := s.curNotifyChan
+		blCh := s.curBlockChan
+		notaryCh := s.curNotaryChan
+		s.RUnlock()
+		select {
+		case <-ctx.Done():
+			break routeloop
+		case ev, ok := <-notifCh:
+			if ok {
+				s.notifyChan <- ev
+			} else {
+				connLost = true
+			}
+		case ev, ok := <-blCh:
+			if ok {
+				s.blockChan <- ev
+			} else {
+				connLost = true
+			}
+		case ev, ok := <-notaryCh:
+			if ok {
+				s.notaryChan <- ev
+			} else {
+				connLost = true
+			}
+		case _, ok := <-cliCh:
+			connLost = !ok
+		case ok := <-restoreCh:
+			restoreInProgress = false
+			if !ok {
+				connLost = true
+			}
+		}
+		if connLost {
+			if !restoreInProgress {
+				s.log.Info("RPC connection lost, attempting reconnect")
+				if !s.client.SwitchRPC() {
+					s.log.Error("can't switch RPC node")
+					break routeloop
+				}
+				cliCh = s.client.NotificationChannel()
+
+				s.Lock()
+				s.curNotifyChan = make(chan *state.ContainedNotificationEvent)
+				s.curBlockChan = make(chan *block.Block)
+				s.curNotaryChan = make(chan *result.NotaryRequestEvent)
+				go s.restoreSubscriptions(s.curNotifyChan, s.curBlockChan, s.curNotaryChan, restoreCh)
+				s.Unlock()
+				restoreInProgress = true
+			drainloop:
+				for {
+					select {
+					case _, ok := <-notifCh:
+						if !ok {
+							notifCh = nil
+						}
+					case _, ok := <-blCh:
+						if !ok {
+							blCh = nil
+						}
+					case _, ok := <-notaryCh:
+						if !ok {
+							notaryCh = nil
+						}
+					default:
+						break drainloop
+					}
+				}
+			} else { // Avoid getting additional !ok events.
+				s.Lock()
+				s.curNotifyChan = nil
+				s.curBlockChan = nil
+				s.curNotaryChan = nil
+				s.Unlock()
+			}
+		}
+	}
+	close(s.notifyChan)
+	close(s.blockChan)
+	close(s.notaryChan)
+}
+
+// restoreSubscriptions restores subscriptions according to
+// cached information about them.
+func (s *subscriber) restoreSubscriptions(notifCh chan<- *state.ContainedNotificationEvent,
+	blCh chan<- *block.Block, notaryCh chan<- *result.NotaryRequestEvent, resCh chan<- bool) {
+	var err error
+
+	// new block events restoration
+	if s.subscribedToNewBlocks {
+		_, err = s.client.ReceiveBlocks(blCh)
+		if err != nil {
+			s.log.Error("could not restore block subscription",
+				zap.Error(err),
+			)
+			resCh <- false
+			return
+		}
+	}
+
+	// notification events restoration
+	for contract := range s.subscribedEvents {
+		contract := contract // See https://github.com/nspcc-dev/neo-go/issues/2890
```

Comment:
But why do we need it? We copy all the provided filters inside neo-go library, see https://github.com/nspcc-dev/neo-go/blob/7b109586ca75e50cdd56a76a68955a7506e9eda3/pkg/neorpc/filters.go#L82.

Reply:
That's #2221, but we still have 0.100.1 used in mainnet (RPC nodes).